giskard/llm/talk/tools/predict.py

```diff
@@ -0,0 +1,153 @@
+                        }
+                    },
+                    "required": ["features_dict"],
+                },
+            },
+        }
+
+    def _get_input_from_dataset(self, row_filter: dict) -> pd.DataFrame:
+        """Get input from dataset.
+
+        Filter rows from the dataset, using the `row_filter`.
+
+        Parameters
+        ----------
+        row_filter : dict
+            The dictionary with features and related values to filter the dataset.
+
+        Returns
+        -------
+        pd.DataFrame
+            The DataFrame with filtered rows.
+        """
+        threshold = 85
+        filtered_df = self._dataset.df.copy()
+
+        for col_name, col_value in list(row_filter.items()):
+            if filtered_df[col_name].dtype == "object":
+                filtered_df = filtered_df[
+                    filtered_df[col_name].apply(lambda x: fuzz.ratio(x.lower(), col_value.lower()) >= threshold)
+                ]
+                if not len(filtered_df):
+                    break
+            else:
+                filtered_df = filtered_df[filtered_df[col_name] == col_value]
+
+        return filtered_df
+
+    def _get_input_from_user(self, feature_values: dict) -> pd.DataFrame:
+        """Form input based on the user's query.
+
+        Take input feature values from user query. For the missed values, take the feature mean from the Dataset.
+
+        Parameters
+        ----------
+        feature_values : dict
+            The dictionary with features and related values extracted from the query.
+
+        Returns
+        -------
+        pd.DataFrame
+            The DataFrame with the formed input.
+        """
+        # Prepare background sample.
+        from giskard.models.model_explanation import _get_background_example
+
+        background_df = self._model.prepare_dataframe(
+            self._dataset.df, self._dataset.column_dtypes, self._dataset.target
+        )
+        background_sample = _get_background_example(background_df, self._dataset.column_types)
+
+        # Fill background sample with known values.
+        for col_name, col_value in list(feature_values.items()):
+            background_sample.loc[0, col_name] = col_value
+
+        return background_sample
+
+    def _prepare_input(self, feature_values: dict) -> Dataset:
+        """Prepare model input.
+
+        Either extract rows form the dataset, or create input from the user's query.
+
+        Parameters
+        ----------
+        feature_values : dict
+            The dictionary with features and related values extracted from the query.
+
+        Returns
+        -------
+        Dataset
+            The Giskard Dataset with model input.
+        """
+        final_input = self._get_input_from_dataset(feature_values)
+
+        # If no data were filtered from the dataset, build vector from the user input.
+        if len(final_input) == 0:
+            final_input = self._get_input_from_user(feature_values)
```

Comment:
Not sure I understand what we are doing here, but shouldn't the input from user always be appended to dataset ?

Reply:
Input from user is never appended to the dataset. It is used to: 1) select rows from the dataset; 2) If no rows were selected, use them as a part of final input row, where the rest of features are just their mean from the dataset.